[spring-cloud-config-client] Authentication not added when spring-cloud-config.url defines a endpoint, not a base url

**Description**
When defining a config with a full endpoint url like https://my-domain-url.com/config
Quarkus does not succeed in getting configuration when it's protected by Basic Authentication 

**Expected behavior**
Successful config retrieve

**Actual behavior** : 401 http response
```
Caused by: java.lang.RuntimeException: Got unexpected HTTP response code 401 from https://my-domain-url.com/config/config/my-app-name/dev
        at io.quarkus.spring.cloud.config.client.runtime.DefaultSpringCloudConfigClientGateway.exchange(DefaultSpringCloudConfigClientGateway.java:66)
        at io.quarkus.spring.cloud.config.client.runtime.SpringCloudConfigServerClientConfigSourceProvider.getConfigSources(SpringCloudConfigServerClientConfigSourceProvider.java:37)
```


**Configuration**
```properties
quarkus.application.name=my-app-name
quarkus.spring-cloud-config.username=user
quarkus.spring-cloud-config.password=pwd
quarkus.spring-cloud-config.fail-fast=true
quarkus.spring-cloud-config.enabled=true
quarkus.spring-cloud-config.url=https://my-domain-url.com/config
```
**Environment :**
 - java version: 1.8.0_181 
 - Quarkus version: 1.4.2 (tried alson in 1.5.2)
 - Build tool : Gradle , version 6.4.1

**Step by step :**
To help resolve the issue , at start-up : 
1) DefaultSpringCloudConfigClientGateway#determineBaseUri line 40 : 
baseUri is the full endpoint "https://my-domain-url.com/config" instead of base uri "https://my-domain-url.com"

2) DefaultSpringCloudConfigClientGateway#setupContext line 83 
``` 
 authCache.put(HttpHost.create(baseUri), new BasicScheme());
 ```
 3)  org.apache.http.client.protocol.RequestAuthCache#process line 53
 Target host is https://my-domain-url.com, port is < 0
 
line 58 :
 ```
 target = new HttpHost(target.getHostName(), route.getTargetHost().getPort(), target.getSchemeName());
```
At this point target is https://my-domain-url.com:443
line 63 : 
```
 AuthScheme authScheme = authCache.get(target);
```
Result : AuthScheme is null, and function RequestAuthCache#doPreemptiveAuth is never call to set Authentication Basic

--- a/extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/DefaultSpringCloudConfigClientGateway.java
+++ b/extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/DefaultSpringCloudConfigClientGateway.java
@@ -1,6 +1,9 @@
 package io.quarkus.spring.cloud.config.client.runtime;
 
-import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
@@ -15,6 +18,7 @@ import org.apache.http.client.config.RequestConfig;
 import org.apache.http.client.methods.CloseableHttpResponse;
 import org.apache.http.client.methods.HttpGet;
 import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.client.utils.URIBuilder;
 import org.apache.http.impl.auth.BasicScheme;
 import org.apache.http.impl.client.BasicCredentialsProvider;
 import org.apache.http.impl.client.CloseableHttpClient;
@@ -30,45 +34,51 @@ class DefaultSpringCloudConfigClientGateway implements SpringCloudConfigClientGa
             .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
 
     private final SpringCloudConfigClientConfig springCloudConfigClientConfig;
-    private final String baseUri;
+    private final URI baseURI;
 
     public DefaultSpringCloudConfigClientGateway(SpringCloudConfigClientConfig springCloudConfigClientConfig) {
         this.springCloudConfigClientConfig = springCloudConfigClientConfig;
-        this.baseUri = determineBaseUri(springCloudConfigClientConfig);
+        try {
+            this.baseURI = determineBaseUri(springCloudConfigClientConfig);
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("Value: '" + springCloudConfigClientConfig.url
+                    + "' of property 'quarkus.spring-cloud-config.url' is invalid", e);
+        }
     }
 
-    private String determineBaseUri(SpringCloudConfigClientConfig springCloudConfigClientConfig) {
-        String baseUri = springCloudConfigClientConfig.url;
-        if (null == baseUri || baseUri.isEmpty()) {
-            throw new IllegalArgumentException("baseUri cannot be empty");
+    private URI determineBaseUri(SpringCloudConfigClientConfig springCloudConfigClientConfig) throws URISyntaxException {
+        String url = springCloudConfigClientConfig.url;
+        if (null == url || url.isEmpty()) {
+            throw new IllegalArgumentException(
+                    "The 'quarkus.spring-cloud-config.url' property cannot be empty");
         }
-        if (baseUri.endsWith("/")) {
-            return baseUri.substring(0, baseUri.length() - 1);
+        if (url.endsWith("/")) {
+            return new URI(url.substring(0, url.length() - 1));
         }
-        return baseUri;
+        return new URI(url);
     }
 
     @Override
-    public Response exchange(String applicationName, String profile) throws IOException {
+    public Response exchange(String applicationName, String profile) throws Exception {
         final RequestConfig requestConfig = RequestConfig.custom()
                 .setConnectionRequestTimeout((int) springCloudConfigClientConfig.connectionTimeout.toMillis())
                 .setSocketTimeout((int) springCloudConfigClientConfig.readTimeout.toMillis())
                 .build();
         final HttpClientBuilder httpClientBuilder = HttpClientBuilder.create().setDefaultRequestConfig(requestConfig);
         try (CloseableHttpClient client = httpClientBuilder.build()) {
-            final String finalUri = baseUri + "/" + applicationName + "/" + profile;
-            final HttpGet request = new HttpGet(finalUri);
+            final URI finalURI = finalURI(applicationName, profile);
+            final HttpGet request = new HttpGet(finalURI);
             request.addHeader("Accept", "application/json");
 
-            HttpClientContext context = setupContext();
+            HttpClientContext context = setupContext(finalURI);
             try (CloseableHttpResponse response = client.execute(request, context)) {
                 if (response.getStatusLine().getStatusCode() != 200) {
                     throw new RuntimeException("Got unexpected HTTP response code " + response.getStatusLine().getStatusCode()
-                            + " from " + finalUri);
+                            + " from " + finalURI);
                 }
                 final HttpEntity entity = response.getEntity();
                 if (entity == null) {
-                    throw new RuntimeException("Got empty HTTP response body " + finalUri);
+                    throw new RuntimeException("Got empty HTTP response body " + finalURI);
                 }
 
                 return OBJECT_MAPPER.readValue(EntityUtils.toString(entity), Response.class);
@@ -76,11 +86,24 @@ class DefaultSpringCloudConfigClientGateway implements SpringCloudConfigClientGa
         }
     }
 
-    private HttpClientContext setupContext() {
+    private URI finalURI(String applicationName, String profile) throws URISyntaxException {
+        URIBuilder result = new URIBuilder(baseURI);
+        if (result.getPort() == -1) {
+            // we need to set the port otherwise auth case doesn't match the request
+            result.setPort(result.getScheme().equalsIgnoreCase("http") ? 80 : 443);
+        }
+        List<String> finalPathSegments = new ArrayList<>(result.getPathSegments());
+        finalPathSegments.add(applicationName);
+        finalPathSegments.add(profile);
+        result.setPathSegments(finalPathSegments);
+        return result.build();
+    }
+
+    private HttpClientContext setupContext(URI finalURI) {
         final HttpClientContext context = HttpClientContext.create();
-        if (baseUri.contains("@") || springCloudConfigClientConfig.usernameAndPasswordSet()) {
+        if ((baseURI.getUserInfo() != null) || springCloudConfigClientConfig.usernameAndPasswordSet()) {
             final AuthCache authCache = InMemoryAuthCache.INSTANCE;
-            authCache.put(HttpHost.create(baseUri), new BasicScheme());
+            authCache.put(new HttpHost(finalURI.getHost(), finalURI.getPort(), finalURI.getScheme()), new BasicScheme());
             context.setAuthCache(authCache);
             if (springCloudConfigClientConfig.usernameAndPasswordSet()) {
                 final CredentialsProvider provider = new BasicCredentialsProvider();

--- a/extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGateway.java
+++ b/extensions/spring-cloud-config-client/runtime/src/main/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGateway.java
@@ -1,8 +1,6 @@
 package io.quarkus.spring.cloud.config.client.runtime;
 
-import java.io.IOException;
-
 interface SpringCloudConfigClientGateway {
 
-    Response exchange(String applicationName, String profile) throws IOException;
+    Response exchange(String applicationName, String profile) throws Exception;
 }

--- a/extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java
+++ b/extensions/spring-cloud-config-client/runtime/src/test/java/io/quarkus/spring/cloud/config/client/runtime/SpringCloudConfigClientGatewayTest.java
@@ -35,7 +35,7 @@ class SpringCloudConfigClientGatewayTest {
     }
 
     @Test
-    void testBasicExchange() throws IOException {
+    void testBasicExchange() throws Exception {
         final String applicationName = "foo";
         final String profile = "dev";
         wireMockServer.stubFor(WireMock.get(String.format("/%s/%s", applicationName, profile)).willReturn(WireMock

--- a/integration-tests/spring-cloud-config-client/src/main/resources/application.yaml
+++ b/integration-tests/spring-cloud-config-client/src/main/resources/application.yaml
@@ -1,7 +1,7 @@
 '%prod':
    quarkus:
      spring-cloud-config:
-       url: http://localhost:8089
+       url: http://localhost:8089/base
        username: user
        password: pass
        enabled: true     

--- a/integration-tests/spring-cloud-config-client/src/main/resources/mappings/a-bootiful-client-prod.json
+++ b/integration-tests/spring-cloud-config-client/src/main/resources/mappings/a-bootiful-client-prod.json
@@ -1,7 +1,7 @@
 {
   "request": {
     "method": "GET",
-    "urlPattern": "/a-bootiful-client/prod",
+    "urlPattern": "/base/a-bootiful-client/prod",
     "basicAuth" : {
       "username" : "user",
       "password" : "pass"

